File: e2e/scenarios/bignum.py
# Arbitrary-precision arithmetic against Python's own `int`.
#
# Python's `int` is arbitrary-precision, so it *is* this type — a free, exact,
# independently written oracle that needs no reference implementation in the
# scenario. That is the same reason the set scenarios live here with Python's
# `set` as their oracle.
#
# These verbs are all prefixed because the names a caller reaches for first --
# pow, divmod, int, abs, min, max, len -- are Python builtins, and monty resolves
# builtins without ever asking the host. A verb named `pow` would silently
# compute Python's own `pow(a, e, m)`, which is exactly the oracle below, so the
# scenario would pass while testing nothing.
#
# Not the kernel oracle. yesno-core's tests/bignum_oracle.rs compares against
# num-bigint over boundary-biased randomized input, which is strictly stronger.
# What is here is the bridge and the boundary.

# Values chosen to cross the interesting widths: below a limb, exactly a limb,
# several limbs, and past the Karatsuba crossover ( 20 limbs = 1280 bits ).
SMALL = 12345
ONE_LIMB = (1 << 64) - 1
TWO_LIMB = (1 << 128) - 3
WIDE = (7 << 1600) | (1 << 800) | 0xDEADBEEF
ALL_ONES = (1 << 1536) - 1

VALUES = [0, 1, SMALL, ONE_LIMB, TWO_LIMB, WIDE, ALL_ONES]

# --- the bridge carries every magnitude, in both directions ------------------
# Expected bit lengths are written out rather than computed. monty has no
# int.bit_length(), and a Python loop that recomputes it would be reimplementing
# the thing under test -- which is what got and_shape.py moved out of scenarios/.
BIT_LENGTHS = [0, 1, 14, 64, 128, 1603, 1536]
LIMB_LENGTHS = [0, 1, 1, 1, 2, 26, 24]

for i in range(len(VALUES)):
    v = VALUES[i]
    assert bn_add(v, 0) == v, v
    assert bn_bit_len(v) == BIT_LENGTHS[i], i
    assert bn_limb_len(v) == LIMB_LENGTHS[i], i
    # A limb is 64 bits, so the two must agree by construction.
    assert bn_limb_len(v) == (bn_bit_len(v) + 63) // 64, i

# Hex is exact and has no leading zeros; the low limb keeps its full 16 digits.
assert bn_hex(0) == "0"
assert bn_hex(255) == "ff"
assert bn_hex(1 << 64) == "10000000000000000"
assert bn_hex(ONE_LIMB) == "ffffffffffffffff"

# --- arithmetic agrees with Python ------------------------------------------
for a in VALUES:
    for b in VALUES:
        assert bn_add(a, b) == a + b
        assert bn_mul(a, b) == a * b

        if a >= b:
            assert bn_sub(a, b) == a - b
        else:
            # None, not a wrap and not a clamp: the answer is not in the
            # domain of an unsigned type.
            assert bn_sub(a, b) is None

        if b == 0:
            assert bn_divmod(a, b) is None
        else:
            q, r = bn_divmod(a, b)
            assert (q, r) == divmod(a, b)
            # Both halves. A quotient digit one too small still satisfies the
            # first identity, because the excess lands in the remainder.
            assert q * b + r == a
            assert r < b

# --- shifts and truncation ---------------------------------------------------
for a in [SMALL, TWO_LIMB, WIDE, ALL_ONES]:
    for n in [0, 1, 63, 64, 65, 127, 128, 1000]:
        assert bn_shl(a, n) == a << n
        assert bn_shr(a, n) == a >> n
        # truncate is exactly `a mod 2**n`, which is what makes it the only
        # write-side overflow rule that agrees with a narrow read.
        assert bn_truncate(a, n) == a % (1 << n)

# --- modular exponentiation --------------------------------------------------
PRIMES = [97, 65537, 2147483647, 1000000007, (1 << 89) - 1]
for p in PRIMES:
    # Fermat: a^(p-1) == 1 mod p for a not divisible by p.
    for a in [2, 3, 7]:
        if a % p != 0:
            assert bn_pow_mod(a, p - 1, p) == 1, (a, p)
    assert bn_pow_mod(WIDE, 37, p) == pow(WIDE, 37, p)

# An even modulus, which Montgomery could never serve and Barrett must.
assert bn_pow_mod(WIDE, 20, 1 << 64) == pow(WIDE, 20, 1 << 64)
# m == 1 makes every residue zero, including the zero exponent -- the classic
# wrong answer here is 1.
assert bn_pow_mod(WIDE, 0, 1) == 0
assert bn_pow_mod(WIDE, 5, 1) == 0
# A zero modulus has no arithmetic at all.
assert bn_pow_mod(2, 3, 0) is None

# --- the OrdSet boundary, without a database ---------------------------------
WIDTH = 1700
STRIDE = 1700
series = [WIDE, 0, SMALL, ALL_ONES % (1 << WIDTH)]
s = bn_series_build(series, WIDTH, STRIDE)
for k, v in enumerate(series):
    assert bn_series_get(s, k, WIDTH, STRIDE) == v, k

# A zero *below* a non-zero value is still counted; the count reaches the
# highest set ordinal.
assert bn_series_count(s, WIDTH, STRIDE) == len(series)

# Reading at a narrower width is exactly `x mod 2**width`. It falls out of the
# least-significant-bit-first ordering, and it is the identity that makes
# truncation the only write rule agreeing with the reader.
for narrow in [1, 63, 64, 65, 300]:
    assert bn_series_get(s, 0, narrow, STRIDE) == WIDE % (1 << narrow), narrow

# An index nothing was written to reads as zero, not as an error. Absence and
# zero are the same thing in a set.
assert bn_series_get(s, 99, WIDTH, STRIDE) == 0

# The seam: at stride 10000, integer 6 spans bits 60000..70000 and crosses the
# 65536-bit chunk boundary. A reader that stopped at the first chunk passes every
# assertion above and fails this one.
STRADDLE_W = 10000
straddle = [0, 0, 0, 0, 0, 0, (1 << STRADDLE_W) - 1]
t = bn_series_build(straddle, STRADDLE_W, STRADDLE_W)
assert bn_series_get(t, 6, STRADDLE_W, STRADDLE_W) == (1 << STRADDLE_W) - 1
assert bn_series_get(t, 5, STRADDLE_W, STRADDLE_W) == 0
assert bn_series_get(t, 7, STRADDLE_W, STRADDLE_W) == 0

File: yesno-e2e/src/bignum.rs
//! `bn_*`: arbitrary-precision integers, and the `OrdSet` boundary they live at.
//!
//! An `OrdSet` under an `IntLayout` is a series of unsigned big integers. These
//! verbs read one out, operate on it, and put it back — the shape
//! `yesno_core::bignum` is built around, so a scenario exercises the real
//! boundary rather than a convenience wrapper.
//!
//! # What these are for
//!
//! **Python's `int` is arbitrary-precision, so it is exactly this type.** That
//! makes it a free, exact, independently-implemented oracle, and a scenario can
//! write `assert bn_mul(a, b) == a * b` with no reference implementation of its
//! own. Same argument that puts the set scenarios here because Python's `set` is
//! the oracle.
//!
//! The **operational sequence**, which no Rust test covers: build a series,
//! store it in a real `Db`, checkpoint, **reopen**, read it back through a
//! `Snapshot`, do arithmetic, store the result, reopen again. A read on a live
//! `Db` is answered from the memtable and never reaches the store — the blind
//! spot that hid three bugs in `mx_*`.
//!
//! **Not the oracle comparison for the kernels.** `yesno-core`'s own
//! `tests/bignum_oracle.rs` does that better: randomized, boundary-biased, and
//! against `num-bigint`. Reimplementing Karatsuba in Python to compare against
//! another Python walk is what got `and_shape.py` moved out of `scenarios/`.
//!
//! **No `bn_store` / `bn_load`.** `bn_series_build` returns an ordinary set
//! handle, so `batch_store_set` and `snap_load_set` already carry it to and from
//! a database. Adding a `bn_`-prefixed pair would duplicate two shipped verbs
//! and give the family two more names to keep alive — `TESTING.md` records
//! thirty-four verbs orphaned when two scenario files moved out.
//!
//! # Every name a caller would reach for first is a Python builtin
//!
//! `pow`, `divmod`, `int`, `abs`, `min`, `max`, `len` — monty resolves builtins
//! **without ever asking the host**, so a verb named `pow` would silently compute
//! Python's `pow( a, e, m )`. That is *precisely the oracle these verbs are
//! compared against*, so the scenario would pass while testing nothing at all.
//! This is the sharpest instance of the collision rule in the harness, and the
//! `bn_` prefix is what makes it a non-issue.

use monty_types::{MontyException, MontyObject};
use yesno_core::bignum::{IntLayout, IntSink};

use crate::convert::{bignum_obj, int_obj, tuple, value_err, Args};
use crate::world::World;

pub const OWNS: &[&str] = &[
    // arithmetic
    "bn_add",
    "bn_sub",
    "bn_mul",
    "bn_divmod",
    "bn_pow_mod",
    "bn_shl",
    "bn_shr",
    "bn_truncate",
    // inspection
    "bn_bit_len",
    "bn_limb_len",
    "bn_hex",
    // the OrdSet boundary
    "bn_series_build",
    "bn_series_get",
    "bn_series_count",
];

/// The layout every series verb interprets its index under.
///
/// Spelled from two explicit arguments rather than carried on a handle, so a
/// scenario that reads at a different width than it wrote says so on the line
/// that does it — which is the `x mod 2^width` identity `read_int` exists to
/// expose.
fn layout_of(verb: &str, width: u64, stride: u64) -> Result<IntLayout, MontyException> {
    let width_bits = u32::try_from(width)
        .map_err(|_| value_err(format!("{verb}(): width_bits must fit a u32")))?;
    let l = IntLayout { width_bits, stride };
    l.check().map_err(|e| value_err(format!("{verb}(): {e}")))?;
    Ok(l)
}

impl World {
    pub(crate) fn call_bignum(
        &mut self,
        verb: &str,
        a: &Args,
    ) -> Result<MontyObject, MontyException> {
        a.no_kwargs()?;
        match verb {
            "bn_add" => {
                a.exact(2)?;
                Ok(bignum_obj(&a.bignum(0)?.add(&a.bignum(1)?)))
            }
            // `None` rather than an exception on underflow, because that is
            // what `BigUint::sub` returns and the point is that it is neither a
            // wrap nor a clamp. A scenario asserts `bn_sub(b, a) is None`.
            "bn_sub" => {
                a.exact(2)?;
                Ok(match a.bignum(0)?.sub(&a.bignum(1)?) {
                    Some(d) => bignum_obj(&d),
                    None => MontyObject::None,
                })
            }
            "bn_mul" => {
                a.exact(2)?;
                Ok(bignum_obj(&a.bignum(0)?.mul(&a.bignum(1)?)))
            }
            // `None` for a zero divisor, matching `divrem`. Not a raise: a
            // zero read out of a sparse series is ordinary data.
            "bn_divmod" => {
                a.exact(2)?;
                Ok(match a.bignum(0)?.divrem(&a.bignum(1)?) {
                    Some((q, r)) => tuple(vec![bignum_obj(&q), bignum_obj(&r)]),
                    None => MontyObject::None,
                })
            }
            "bn_pow_mod" => {
                a.exact(3)?;
                Ok(match a.bignum(0)?.pow_mod(&a.bignum(1)?, &a.bignum(2)?) {
                    Some(v) => bignum_obj(&v),
                    None => MontyObject::None,
                })
            }
            "bn_shl" => {
                a.exact(2)?;
                Ok(bignum_obj(&a.bignum(0)?.shl(a.u64(1)?)))
            }
            "bn_shr" => {
                a.exact(2)?;
                Ok(bignum_obj(&a.bignum(0)?.shr(a.u64(1)?)))
            }
            "bn_truncate" => {
                a.exact(2)?;
                Ok(bignum_obj(&a.bignum(0)?.truncate(a.u64(1)?)))
            }
            "bn_bit_len" => {
                a.exact(1)?;
                Ok(int_obj(a.bignum(0)?.bit_len()))
            }
            "bn_limb_len" => {
                a.exact(1)?;
                Ok(int_obj(a.bignum(0)?.limbs().len() as u64))
            }
            "bn_hex" => {
                a.exact(1)?;
                Ok(MontyObject::String(a.bignum(0)?.to_hex_string()))
            }

            // --- the OrdSet boundary ---
            // Returns an ordinary set handle, so the database verbs carry it.
            "bn_series_build" => {
                a.exact(3)?;
                let layout = layout_of(verb, a.u64(1)?, a.u64(2)?)?;
                let values = a.bignum_list(0)?;
                let mut sink = IntSink::new(layout);
                for (k, v) in values.iter().enumerate() {
                    sink.place(k as u64, v)
                        .map_err(|e| value_err(format!("{verb}(): index {k}: {e}")))?;
                }
                Ok(self.push_set(sink.build()))
            }
            "bn_series_get" => {
                a.exact(4)?;
                let set = self.set(a.handle(0)?, verb)?.clone();
                let layout = layout_of(verb, a.u64(2)?, a.u64(3)?)?;
                Ok(match set.read_int(a.u64(1)?, &layout) {
                    Some(v) => bignum_obj(&v),
                    // Not addressable at all, which is a statement about the
                    // layout rather than an empty answer.
                    None => MontyObject::None,
                })
            }
            "bn_series_count" => {
                a.exact(3)?;
                let set = self.set(a.handle(0)?, verb)?.clone();
                let layout = layout_of(verb, a.u64(1)?, a.u64(2)?)?;
                Ok(int_obj(set.int_count(&layout)))
            }
            _ => Err(value_err(format!("{verb}() is not a bignum verb"))),
        }
    }
}
